### ISSUE TITLE
Dockerize Solr for integration tests

### ISSUE THREAD

**Issue by gjunge** [SolrNet/SolrNet#299]
Instead of having a copy of Solr sitting inside the repository, use the public [Docker image](https://store.docker.com/images/solr)

- We can add cores/data on startup, and do store this config in the repository. This means we can easily clean up Solr for each test.
- We can even have docker start when the tests start (http://btburnett.com/2017/02/restful-api-integration-testing-for-net-core-using-docker.html)
- Upgrading Solr is easy

**Comment by xavier-morera** [comment]
@gjunge Definitively a great idea. I am still hanging on to your other suggestion of using PowerShell to download Solr, extract and setup. It could be configured to use 4.x, 5.x or 6.x.

We may want to explore both options and decide. I do like Docker as well, just thinking of what would make it easier for everyone else. 

What do you think? Is the Powershell option worth exploring too?

**Comment by gjunge** [comment]
@xavier-morera The powershell script was my first thought, after having downloaded the Windows version of Solr 6.6 myself and trying to set it up. Afterwards I thought of checking whether there is a docker version available. 

Both have their pros and cons, I'm inclined towards docker due to its seemingly simplicity  (haven't tried It yet though). So probably the list below is biased. 

PS pros:
- No need to install Docker. 
- Can add support to as many SOLR versions wanted. 

PS cons:
- it doesn't just work on every computer (for instance I had to tweak the start SOLR script to make it look at the right registry setting to find java)
- dependencies on java
- cannot easily switch between versions of Solr
- statefull solution, have to manually clean up. 
- another script to maintain

Docker pros:
- besides maintaining instance setup no need to maintain setup code
- easily switch SOLR versions (5.5 and up)
- stateless, always get a clean slate
- it's a container so just works

Docker cons:
- dependency on Docker
- only get he versions that are in the docker library. 

In theory it should also be easier in docker to create multiple environments such as a  SolrCloud setup and single Solr setup, just depending on a different composition. In PS it should also be possible but takes much more scripting which creates more bugs which are not the type of issues contributors should deal with, rather work on SolrNet itself.

**Comment by xavier-morera** [comment]
@gjunge good points on Docker. Seems like it is a better alternative than Powershell, mainly to reduce dependencies and for simplicity. Also, anyone who is using a particular version could deploy their specific version and use it as well.

**Comment by serbrech** [comment]
+1 for docker
Installing Java and Solr permanently on my local machine, to be able to develop is not a good trade-off.
this alone will make me go for docker any day. That's what docker is for.
Docker is embraced by windows today, with the Linux Subsystem, so I do not see this as an anoying dependency. It's part of the everyday dev toolbelt.
I could potentially help with setting this up (I've done it before).

**Comment by xavier-morera** [comment]
@serbrech agreed. It is what I just mentioned after reading @gjunge 's comments. 

If you can help, that would be great.

**Comment by gjunge** [comment]
@serbrech  That would indeed be great!

The Docker config should tailor our integration tests, so with other words should enable good integration tests.

What I've seen being done in the integration tests:

* Use Solr with a single core, and a quit extensive schema (with all kind of datatypes)
* Use Solr with multiple cores (testing the multi core features, so datatypes are not so important)
* Adding cores
* Removing cores
* Swapping cores
* Aliasing cores

And we will have to add integration tests surrounding SolrCloud in the near future.

As a side note, this Solr setup was also in use for the SampleApp (which has been moved to its own repo: https://github.com/SolrNet/SolrNetSampleApps), so if we could reuse the docker setup for that as well, that would be great.

**Comment by xavier-morera** [comment]
@gjunge and @serbrech I decided to play around with Docker with Solr today and it is pretty easy and works pretty well. I did a few tests, here are some of the things that I learned today (being honest, my first time working with Docker - I use VirtualBox and other virtualization solutions a lot, but Docker not so much - until probably now moving forward)
- [Docker](https://docs.docker.com/docker-for-windows/install/) on Windows runs only with Windows 10 pro or Server 2016 RTM 
- For other versions of Windows you need [Docker Toolbox](https://docs.docker.com/toolbox/toolbox_install_windows/). I ran into an issue as I was already running VirtualBox so I had to manually reinstall the VirtualBox driver, not a big deal and I can document the three steps required.  
- There are [multiple versions of Solr available](https://hub.docker.com/_/solr/), and there is even a version with a demo core with sample data
![image](https://user-images.githubusercontent.com/7174278/33237547-ec594442-d23b-11e7-84f4-d33736b9296d.png)

Starting a server is absurdly easy, just takes a bit to download but you run this command
`docker run --name my_solr -d -p 8983:8983 -t solr`

- You wait for a bit to download
![image](https://user-images.githubusercontent.com/7174278/33237540-b7301f16-d23b-11e7-81fa-be48f2df3daa.png)

- And you get this
![image](https://user-images.githubusercontent.com/7174278/33237552-24494c94-d23c-11e7-95f6-9b29527aee1f.png)

Change the tags and you get other versions. Please note 5.5 is the minimum version in the "official" list.

- I also installed on Mac and works well

- Additionally I tested downloading the zip file, expanding and starting to have an alternative to Docker but there are a few dependencies and prerequisites that could make it hard to go this route for integration tests

- And finally, I gave it a shot at running Docker within a VM. I assumed it wouldn't work, and indeed that was the case. I know this sounds a bit dumb, but I virtualize all my dev stuff, so I had to give it a shot.

All in all, I think Docker seems to be the best way to get multiple versions of Solr up and running instead of having to upload to Git a version of Solr.

Thoughts?

**Comment by gjunge** [comment]
@xavier-morera Docker is really great, certainly for these scenarios. For testing you just want everything to work in a lab-like environment, with no external interference, and Docker gives us that with the added simplicity.

I also believe that Docker becomes more and more a standard tool in each developers' toolkit, so the Docker requirement is not that outrageous.

Dockerizing our tests, and our sample app is one of the most important things before moving forward, it will make adding features so much "safer".

I see the following pain/attention points:

* Our current tests work in general with one set of data/schema. I think we should settle on a standard schema setup that would cover most of our tests. You said that some of the Docker images come with sample data, would be great to use that , although I'm not sure if we can rely on it to be the same throughout all the images. Because we wouldn't want our tests suddenly to fail because they decided to change the Sample data.
* Rewrite our tests after we settle on the sample data.
* Our Docker setup, should include both Solr and Solr Cloud support.
* Nice to have: Have AppVeyor  also run the integration tests. (It does have Docker enabled: https://www.appveyor.com/docs/build-environment/#services)
* Nice to have: have our integration test iterate a few version of Solr , so we automatically run the tests on a number of different Docker images.
* Rewrite our Sample App to work with this sample data.

**Comment by xavier-morera** [comment]
I included a page on how to set up Docker and run a container with Solr 5.5.5 to 7.1. Will expand with more info on using Docker.

Responding to your questions, @gjunge:
* Our current tests work in general with one set of data/schema. I think we should settle on a standard schema setup that would cover most of our tests. You said that some of the Docker images come with sample data, would be great to use that , although I'm not sure if we can rely on it to be the same throughout all the images. Because we wouldn't want our tests suddenly to fail because they decided to change the Sample data.

*Agreed*

* Rewrite our tests after we settle on the sample data.

*Agreed*

* Our Docker setup, should include both Solr and Solr Cloud support.

*Agreed. I posted instructions that use one set of images, there are others that include SolrCloud as well. We just need to find one that's reliable, stable and well configured.*

* Nice to have: Have AppVeyor  also run the integration tests. (It does have Docker enabled: https://www.appveyor.com/docs/build-environment/#services)

*This will be great*

* Nice to have: have our integration test iterate a few version of Solr , so we automatically run the tests on a number of different Docker images.

*Shouldn't be hard, a few approaches could be* 
- *Start one container at a time, run the tests, stop container, start new container with different solr version, test, stop container, ...*
- *Or allow multiple containers to be up and running at a time to allow for testing in parallel against multiple versions. This would require a bit more hardware, but can make life simpler. *

*Something good is that this will make it a lot simpler to find breaking changes in Solr, deprecated functionality and other issues that arise from different versions*

* Rewrite our Sample App to work with this sample data.

*The sample app is nice, but it is really really old. However, writing a new full front end could be too time consuming - time that could be spent on features in the core of the project. Unless we can get someone with front end development skills to chip in.*

**Comment by gjunge** [comment]
@xavier-morera - excellent.

I think this issue now contains all our requirements, and a guideline on how to implement.
So the next step is to get somebody to pick this up, and then we can continue to discuss the details if necessary.
For the coming few weeks I'm a little busy, so I will not be able to take this one for the coming while.

**Comment by xavier-morera** [comment]
@gjunge thanks.

**Comment by mausch** [comment]
> Instead of having a copy of Solr sitting inside the repository, use the public [Docker image](https://store.docker.com/images/solr)

Did this in 65d74d66273be7dd993154399798b75d3e856b62
Still using the "techproducts" schema that comes built into Solr.
The integration tests do include SolrCloud.

> Nice to have: Have AppVeyor also run the integration tests.

They run on Github Actions.

> Nice to have: have our integration test iterate a few version of Solr , so we automatically run the tests on a number of different Docker images.

It currently runs on each major version of Solr from 5.x to 8.x
Solr 9 was released a couple of months ago, should add it too.

> The sample app is nice, but it is really really old. However, writing a new full front end could be too time consuming 

[I upgraded the tech in the sample app about a year ago](https://github.com/SolrNet/SolrNetSampleApps/commit/1948befa40c8c13e604db5b031dea44c9ebb98ed), though I didn't really change any of its features.
